statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,698 @@
+
+    Parameters
+    ----------
+    mod : statsmodels model class instance
+        The model for the current partition.
+    params : array-like
+        The estimated coefficients for the current partition.
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+    L1_wt : scalar
+        The fraction of the penalty given to the L1 penalty term.
+        Must be between 0 and 1 (inclusive).  If 0, the fit is
+        a ridge fit, if 1 it is a lasso fit.
+    score_kwds : dict-like or None
+        Keyword arguments for the score function.
+
+    Returns
+    -------
+    An array-like object of the same dimension as params
+
+    Notes
+    -----
+    In general:
+
+    gradient l_k(params)
+
+    where k corresponds to the index of the partition
+
+    For the simple linear case:
+
+    X^T(y - X^T params)
+    """
+
+    grad = -mod.score(np.r_[params], **score_kwds)
+    grad += alpha * (1 - L1_wt)
+    return grad
+
+
+def _calc_wdesign_mat(mod, params, hess_kwds):
+    """calculates the weighted design matrix necessary to generate
+    the approximate inverse covariance matrix
+
+    Parameters
+    ----------
+    mod : statsmodels model class instance
+        The model for the current partition.
+    params : array-like
+        The estimated coefficients for the current partition.
+    hess_kwds : dict-like or None
+        Keyword arguments for the hessian function.
+
+    Returns
+    -------
+    An array-like object, updated design matrix, same dimension
+    as mod.exog
+    """
+
+    # TODO correctly handle hessian_obs
+    rhess = np.sqrt(mod.hessian_obs(np.asarray(params), **hess_kwds))
+    return rhess[:, None] * mod.exog
+
+
+def _calc_nodewise_row(wexog, idx, alpha):
+    """calculates the nodewise_row values for the idxth variable, used to
+    estimate approx_inv_cov.
+
+    Parameters
+    ----------
+    wexog : array-like
+        The weighted design matrix for the current partition.
+    idx : scalar
+        Index of the current variable.
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+
+    Returns
+    -------
+    An array-like object of length p-1
+
+    Notes
+    -----
+
+    nodewise_row_i = arg min 1/(2n) ||wexog_i - wexog_-i gamma||_2^2
+                             + alpha ||gamma||_1
+    """
+
+    p = wexog.shape[1]
+    # handle array alphas
+    if not np.isscalar(alpha):
+        alpha = alpha[ind]
+
+    ind = list(range(p))
+    ind.pop(idx)
+
+    tmod = OLS(wexog[:, idx], wexog[:, ind])
+
+    nodewise_row = tmod.fit_regularized(alpha=alpha).params
+
```

Comment:
To make sure I understand what you are thinking here, since we use `ind` in multiple parts of the nodewise regression do you think we should generate it externally and pass it into each function?